unittests/svnn_ibc_tests.cpp

```diff
@@ -75,66 +75,217 @@ BOOST_AUTO_TEST_SUITE(svnn_ibc)
-      auto genesis_block_action_mroot = genesis_block_fd.value().action_mroot;
+   // The ibc_cluster class inherits from finality_test_cluster and serves to generate finality proofs in the context of IBC.
+
+   // It has its own high level produce_block function, which hides all the internal finality details, and returns an extended struct containing data relevant to IBC.
+
+   // It also tracks the Savanna state in a rudimentary manner, sufficient to generate the proofs we care about.
+
+   // Since the ibc contract only cares about verifying finality proofs, ibc_cluster doesn't support forks or rollbacks, and always assumes the happy path in finality progression.
+
+   // It also assumes a single producer pre-transition, resulting in only 2 transition blocks.
+
+   // *****
+
+   template<size_t NUM_NODES>
+   class ibc_cluster : public finality_test_cluster<NUM_NODES> {
+   public:
+
+      // data relevant to IBC
+      struct ibc_block_data_t {
+         signed_block_ptr block;
+         action_trace onblock_trace;
+         finality_data_t finality_data;
+         digest_type action_mroot;
+         digest_type base_digest;
+         digest_type active_finalizer_policy_digest;
+         digest_type last_pending_finalizer_policy_digest;
+         digest_type last_proposed_finalizer_policy_digest;
+         digest_type finality_digest;
+         digest_type computed_finality_digest;
+         digest_type afp_base_digest;
+         digest_type finality_leaf;
+         digest_type finality_root;
+      };
+
+      // cache last proposed, last pending and currently active finalizer policies + digests
+      eosio::chain::finalizer_policy last_proposed_finalizer_policy;
+      digest_type last_proposed_finalizer_policy_digest;
+
+      eosio::chain::finalizer_policy last_pending_finalizer_policy;
+      digest_type last_pending_finalizer_policy_digest;
+
+      eosio::chain::finalizer_policy active_finalizer_policy;
+      digest_type active_finalizer_policy_digest;
+
+      // counter to (optimistically) track internal policy changes
+      uint32_t blocks_since_proposed_policy = 0;
+
+      // internal flag to indicate whether or not block is the IF genesis block
+      bool is_genesis = true;
+
+      bool is_transition = true;
+
+      // returns finality leaves for construction of merkle proofs
+      std::vector<digest_type> get_finality_leaves(const size_t cutoff){
+         assert(cutoff>=0 && cutoff<finality_leaves.size());
+         return std::vector<digest_type>(finality_leaves.begin(), finality_leaves.begin() + cutoff + 1);
+      }
+
+      // produce and propagate a block, update internal state as needed, and returns relevant IBC data 
+      ibc_block_data_t produce_block(){
+         auto result = this->produce_and_push_block_ex();
+         signed_block_ptr block = result.block;
+         action_trace onblock_trace = result.onblock_trace->action_traces[0];

```

Comment:
It is possible that the `onblock_trace` is nullptr. Should add a BOOST_REQUIRE that it exists.